core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -418,102 +425,103 @@ public HashDistributionTaskSource(
         }
 
         @Override
-        public List<TaskDescriptor> getMoreTasks()
+        public ListenableFuture<List<TaskDescriptor>> getMoreTasks()
         {
-            if (finished || closed) {
-                return ImmutableList.of();
+            synchronized (this) {
+                if (finished || closed) {
+                    return immediateFuture(ImmutableList.of());
+                }
             }
 
             Map<Integer, ListMultimap<PlanNodeId, Split>> partitionToSplitsMap = new HashMap<>();
             SetMultimap<Integer, HostAddress> partitionToNodeMap = HashMultimap.create();
-            for (Map.Entry<PlanNodeId, SplitSource> entry : splitSources.entrySet()) {
-                SplitSource splitSource = entry.getValue();
-                BucketNodeMap bucketNodeMap = this.bucketNodeMap
-                        .orElseThrow(() -> new VerifyException("bucket to node map is expected to be present"));
-                while (!splitSource.isFinished()) {
-                    ListenableFuture<SplitBatch> splitBatchFuture = splitSource.getNextBatch(NOT_PARTITIONED, Lifespan.taskWide(), splitBatchSize);
-
-                    long start = System.nanoTime();
-                    addSuccessCallback(splitBatchFuture, () -> getSplitTimeRecorder.accept(start));
-
-                    SplitBatch splitBatch = getFutureValue(splitBatchFuture);
-
-                    for (Split split : splitBatch.getSplits()) {
-                        int bucket = bucketNodeMap.getBucket(split);
-                        int partition = getPartitionForBucket(bucket);
-
-                        if (!bucketNodeMap.isDynamic()) {
-                            HostAddress requiredAddress = bucketNodeMap.getAssignedNode(split).get().getHostAndPort();
-                            Set<HostAddress> existingRequirement = partitionToNodeMap.get(partition);
-                            if (existingRequirement.isEmpty()) {
-                                existingRequirement.add(requiredAddress);
-                            }
-                            else {
-                                checkState(
-                                        existingRequirement.contains(requiredAddress),
-                                        "Unable to satisfy host requirement for partition %s. Existing requirement %s; Current split requirement: %s;",
-                                        partition,
-                                        existingRequirement,
-                                        requiredAddress);
-                                existingRequirement.removeIf(host -> !host.equals(requiredAddress));
+            List<ListenableFuture<LoadedSplits>> splitSourceCompletionFutures = splitSources.entrySet().stream()
```

Comment:
Maybe also better enforce that it is not called more than once (when there's a future in process?)